feat: Support custom S3 endpoint configuration

dingo4dev opened the pull request:
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

This PR introduces support for a custom S3 endpoint URL

This allows users to connect to any S3-compatible object storage solution (such as MinIO, Ceph, LocalStack, etc.) instead of being limited to the default AWS S3 endpoints.

<!-- Signed-off-by: stanleylaw <19900516+dingo4dev@users.noreply.github.com> -->

dingo4dev committed:
Adds endpoint URL to S3 storage config

Enables specifying an endpoint URL for S3 storage configurations.
This allows connecting to S3-compatible storage services that
may not be hosted on AWS.

The endpoint URL is now included when building the S3 client,
allowing to override the default AWS endpoint.
487f89

Signed-off-by: Stanley <19900516+dingo4dev@users.noreply.github.com>

Replace  S3Client Java 1.x  to 2.x

use aws utils for s3 client and add unit test

add s3 endpoint to iceberg table config

update docs

Signed-off-by: DinGo4Dev <19900516+dingo4dev@users.noreply.github.com>
Signed-off-by: Stanley Law <19900516+dingo4dev@users.noreply.github.com>

RG-sw commented:
@dingo4dev I was neeeding this a lot, thank you! :pray:

I'm trying to test it locally (docker compose setup + minio), I have this `server.properties`
```
server.env=dev

s3.bucketPath.0=s3://warehouse
s3.region.0=eu-west-2
s3.accessKey.0=minio
s3.secretKey.0=minio123
s3.sessionToken.0=test-token
s3.endpointUrl.0=minio:9000
```

Than I create a `test.test2` schema and from CLI I try:

```
bin/uc table create \
  --full_name test.test2.my_table \
  --columns "id INT, name STRING" \
  --storage_location "s3://warehouse" ----> WORKS
```

Then I try to write random data in it:
```
bin/uc table write \
  --full_name test.test2.my_table
```
And I get the error: `Error occurred while executing the command. Failed to write sample data to delta table s3://warehouseFailed to create delta table`


Am I doing something wrong ? How did you test it ?

dingo4dev commented:
@RG-sw Thanks for raising this. After some testing, the custom endpoint is missing in the STS credentials part which I am still WIP. BTW, there is a new PR https://github.com/unitycatalog/unitycatalog/pull/1158 have implement it.

dingo4dev committed:
Adds endpoint URL to AWS STS credential generation

Enables specifying an endpoint URL when generating AWS credentials. This allows the application to connect to S3-compatible storage solutions that are not hosted on AWS.

The endpoint URL is now passed to the `StsCredentialsGenerator` to allow overriding the default AWS endpoint.

Signed-off-by: DinGo4Dev <19900516+dingo4dev@users.noreply.github.com>
Signed-off-by: Stanley Law <19900516+dingo4dev@users.noreply.github.com>

dingo4dev committed:
fix styling of "*" import

Signed-off-by: Stanley Law <19900516+dingo4dev@users.noreply.github.com>

dishkakrauch commented:
Hello. Is this MR still in progress?

dingo4dev commented:
@dishkakrauch Hi
I know things have been a bit quiet around here lately, and since it's been a while since my last update, I've slightly lost track of our current status. It seems like this PR might be blocked.

Whenever you or anyone else on the team has a few open cycles, I'd love to get some guidance on the next steps to get this moving forward.

dishkakrauch commented:
> Whenever you or anyone else on the team has a few open cycles, I'd love to get some guidance on the next steps to get this moving forward.

Hi. Unfortunatelly I'm not a part of unity catalog team but would like to be)
